### PR TITLE
Add MIRI to CI and fix UB

### LINKED ISSUE #62: MIRI reports UB on `<Vec<u8> as ByteVecExt>::try_extend_zeroed`
Running the test suite in this repository through MIRI reports that `<Vec<u8> as ByteVecExt>::try_extend_zeroed` results in Undefined Behaviour. 

I believe that this is because the end of `as_mut_ptr_end` is open and therefore is not pointing to anything borrowed. I'll open a pull request in a minute with a fix which both adds MIRI to CI and which passes all tests with MIRI.

### PR DESCRIPTION
**Pull request by Joeoc2001**
Resolves #62

### CODE REVIEW

**Comment by teoxoy** [comment]
Could you force-push the commits again? CI seems to be stuck and I think that should re-trigger it.

**Comment by Joeoc2001** [comment]
I've tried force pushing, and adding a nop commit, but neither seems to have unstuck CI - is it because I added MIRI to the github actions, which is theoretically insecure to run without approval? Do you need to sign off on my workflow changes somehow before checks will run?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -90,3 +90,16 @@ jobs:
           sudo apt install -y mesa-vulkan-drivers vulkan-validationlayers
 
       - run: cargo llvm-cov nextest --workspace --features mint,glam --fail-under-lines 80
+
+  docs:
```
Comment (teoxoy) [inline .github/workflows/ci.yml:94]:
```suggestion
  tests_miri:
```

I think this is the cause of the CI being stuck (there is already a `docs` job).

**Comment by Joeoc2001** [comment]
Cheers! Silly mistake. 

Now it's failing on unstable library functions due to the old(er) nightly build being used to test - issue 86161 was stabilized July of last year. Since the latest stable release is about a year more recent than the nightly being used in the CI, I'll promote the nightly by a few months to match the latest stable release (1.75.0, 2023-12-28).

File: src/utils.rs [hunk 2]
```diff
@@ -78,14 +78,14 @@ impl ByteVecExt for Vec<u8> {
...
-            let end = self.as_mut_ptr_range().end;
+            let end = unsafe { self.as_mut_ptr().add(self.len()) };
             // SAFETY
             // 1. dst ptr is valid for writes of count * size_of::<T>() bytes since the call to Vec::reserve() succeeded
             // 2. dst ptr is properly aligned since we got it via Vec::as_mut_ptr_range()
             unsafe { end.write_bytes(u8::MIN, additional) }
             // SAFETY
             // 1. new_len is less than or equal to Vec::capacity() since we reserved at least `additional` elements
-            // 2. The elements at old_len..new_len are initialized since we wrote `additional` bytes
+            // 2. The elements at old_len..new_len are initialized since we wrote `additional` bytes.
```
Comment (teoxoy) [inline src/utils.rs:88]:
```suggestion
            // 2. The elements at old_len..new_len are initialized since we wrote `additional` bytes
```

**Comment by teoxoy** [comment]
Sounds good, the versions are pinned just to avoid random breakages.

**Comment by teoxoy** [comment]
It would be nice if GitHub actually told us: "Hey the workflow file is invalid!" in big red text 😅

**Review by teoxoy — APPROVED** [review]
Looks good, thanks!

**Pull request merged**